src/function.jl

```diff
@@ -0,0 +1,19 @@
+abstract type AbstractOptimizationFunction end
+
+struct OptimizationFunction{F,G,H,K} <: AbstractOptimizationFunction
+    f::F
+    grad::G
+    hess::H
+    kwargs::K
+end
+
+function OptimizationFunction(f; grad=nothing,hess=nothing, p=DiffEqBase.NullParameters(),kwargs...)
+    if p isa DiffEqBase.NullParameters
+        g! = (res,x) -> ForwardDiff.gradient!(res, f, x)
+        h! = (res,x) -> ForwardDiff.hessian!(res, f, x)
+    else
+        g! = (res,x) -> ForwardDiff.gradient!(res, θ -> f(θ,p), x)
```

Comment:
we want to calculate the output and the gradient simultaneously. Just copy DiffEqFlux here.